service-registry/src/main/java/org/apache/servicecomb/serviceregistry/consumer/SimpleMicroserviceInstancePing.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.servicecomb.serviceregistry.consumer;
+
+import java.io.IOException;
+import java.net.Socket;
+
+import org.apache.servicecomb.foundation.common.net.IpPort;
+import org.apache.servicecomb.foundation.common.net.NetUtils;
+import org.apache.servicecomb.serviceregistry.api.registry.MicroserviceInstance;
+
+/**
+ * Simple implementation of .MicroserviceInstancePing using telnet
+ */
+public class SimpleMicroserviceInstancePing implements MicroserviceInstancePing {
+  @Override
+  public int getOrder() {
+    return 100;
+  }
+
+  @Override
+  public boolean ping(MicroserviceInstance instance) {
+    if (instance.getEndpoints() != null && instance.getEndpoints().size() > 0) {
+      IpPort ipPort = NetUtils.parseIpPortFromURI(instance.getEndpoints().get(0));
+      try (Socket s = new Socket(ipPort.getHostOrIp(), ipPort.getPort())) {
```

Comment:
lost heartbeat means tens of seconds passed
if deployment managered by docker, in this time, maybe already start another microservce instance for this ip and port

Reply:
I have think this problem very carefully. This mechanism only work when client get empty instances list, and we face this problem due to service center bugs, unexpected start/stop of microservices and service center, and other network failures not properly handled. In theory, these problems should be properly handled and SDK do not need to care about it, but It happens very frequently. 
So we add this mechanism to make SDK itself more stable in bad circumstances. Since SDK will query instances periodically, after there are instances, no problem for your mentioned scenario. And if even your scenario happens(rarely), this time, to consumer both of them are failures(no instance vs send to wrong destination). This time is very short and acceptable.